### ISSUE TITLE
Support ignoring whitespace/indentation differences in fn:deep-equal.

### ISSUE THREAD

**Issue by rhdunn** [qt4cg/qtspecs#98]
## Signatures
```
fn:deep-equal( $input1 as item()*,
               $input2 as item()*,
               $collation as xs:string? := (),
               $boundary-space as enum("preserve", "strip") := "preserve") as xs:boolean
```
## Notes
If `$boundary-space` is "preserve", then any whitespace differences are checked and would result in the function returning `false()`.

If `$boundary-space` is "strip", then any whitespace differences are ignored and would result in the function returning `true()` if the inputs are otherwise identical.
## Use Case
Comparing two XML fragments in a unit test assertion where you don't care about indentation differences.

**Comment by ChristianGruen** [comment]
Good idea.

I believe that both Saxon (?) and BaseX (!) have included support for hidden deep-equal arguments, partially inspired by the requirements of the QT3 Test Suite. Maybe the fourth argument should be a map, such that we could make existing options public. In BaseX, we have flags for:

* skipping the comparison of descendant comments and processing instructions
* comparing namespaces

**Comment by rhdunn** [comment]
I'm happy for the last argument to be a map. We could define a set of required options (e.g. `boundary-space`). but leave it open for vendors to extend.

**Comment by rhdunn** [comment]
Here is an updated proposal. I've not specified any other deep-equal options, but we may wish to do that to at least cover the options needed for the QT3 tests that are supported by BaseX and Saxon.

## Signature
```
fn:deep-equal( $input1 as item()*,
               $input2 as item()*,
               $collation as xs:string? := (),
               $options as item-type(deep-equal-options) := map {
                   "boundary-space": "preserve"
               }) as xs:boolean
```
## Rules
The `$deep-equal-options` parameter is represented as a value of type `deep-equal-options`, defined as follows:
```
record(
    boundary-space as enum("preserve", "strip"),
    *
)
```

If a `text()` node has a length of 0 when `fn:normalize-space` is applied to it, and the `boundary-space` option of `$deep-equal-options` is set to "strip", then that text node is ignored during the comparison.

TODO: Q: Should this be `fn:normalize-space($a-text) eq fn:normalize-space($b-text)` instead? I.e. should `fn:deep-equal(<a> one </a>, <a>one</a>, boundary-space: "strip")` return `true()`?

TODO: Formalize the behaviour for schema aware processors.

The meaning of any additional `$deep-equal-options` is ·implementation-defined·. To avoid conflict with any future version of this specification, the keys of any such entries should start with an underscore character.

**Comment by michaelhkay** [comment]
I'm not keen on the term "boundary space" as it relates specifically to whitespace in queries, not in XML documents (and it takes into account curly braces).

I propose we name options according to some property, and use the value true if a difference in the property is significant, and false if it is insignificant. 

Options and defaults (and meaning):

- namespace-prefixes : false (whether the prefix of an element or attribute name is significant)
- namespace-uris : true (whether the namespace URI of an element name is significant)
- in-scope-namespaces : false (whether the in-scope namespaces of an element are significant)
- whitespace-nodes : true (whether whitespace text nodes are significant)
- whitespace-variation : true (if false, string values are compared after normalize-text)
- unicode-variation : true (if false, string values are compared after normalize-unicode)
- case-variation : true (if false, string values are compared using case-blind collation)
- comments : true (whether comments are significant. If false, the text nodes either side of a comment are merged before the comparison)
- processing-instructions : true (whether PIs are significant, similarly)
- type-annotations : false (whether type annotations are significant; if true, they must match exactly)
- typed-values : false (if true, compare typed values of nodes rather than string values)
- child-order : true (whether the order of child nodes is significant)
- id : false (whether the id property of a node is significant)
- idref : false (whether the idref property of a node is significant)
- nilled : false (whether the nilled property of a node is significant)

I've included the options whitespace-variation, unicode-variation, and case-variation instead of allowing a collation to be used. I simply haven't seen use cases for this kind of function where a collation makes sense.

I propose that there should also be a callback function f(X, Y, M) which is called when two nodes are found to be different; X and Y are the two nodes (or may be empty if one tree has nodes with no correspondent node in the other), and M is an error code (which we will define) indicating the nature of the difference, e.g. STRING-VALUE if the string values of the two nodes differ.

I propose that this function should be for comparing node trees only (not arbitrary sequences, not maps and arrays) and should be called compare-nodes.

**Comment by michaelhkay** [comment]
I believe this issue has been resolved by PR #219.
